File: rust/src/problem.rs
use std::fs::File;
use std::io::BufReader;
use serde::{Deserialize, Serialize};
use serde_json::Result;
use crate::geometry::Point;

#[derive(Deserialize, Debug)]
pub struct Attendee {
    pub x: f64,
    pub y: f64,
    pub tastes: Vec<f64>,
}

#[derive(Deserialize, Debug)]
pub struct Pillar {
    pub center: Vec<f64>,
    pub radius: f64,
}

#[derive(Deserialize, Debug)]
pub struct Problem {
    pub room_width: f64,
    pub room_height: f64,
    pub stage_width: f64,
    pub stage_height: f64,
    pub stage_bottom_left: Vec<f64>,
    pub musicians: Vec<usize>,
    pub attendees: Vec<Attendee>,
    pub pillars: Vec<Pillar>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Solution {
    pub placements: Vec<Point<f64>>,
    pub volumes: Option<Vec<f64>>,
}

impl Problem {

    pub fn from_file(f: &String) -> Result<Problem> {
        let file = File::open(f).unwrap();
        let reader = BufReader::new(file);
        let s: String = serde_json::from_reader(reader)?;
        let p= serde_json::from_str(&*s);
        p
    }

}

impl Solution {
    pub fn from_file(f: &String) -> Result<Solution> {
        let file = File::open(f).unwrap();
        let reader = BufReader::new(file);
        let s = serde_json::from_reader(reader);
        s
    }

    pub fn volume(&self, i: usize) -> f64 {
        match &self.volumes {
            None => 1.0,
            Some(x) => x[i]
        }
    }
}

File: haskell/src/Problem.hs
{-# LANGUAGE OverloadedStrings, NamedFieldPuns #-}
module Problem (Problem(..),
                Attendee(..),
                Pos(..),
                Pillar(..),
                readProblem, readSolution) where

import qualified Data.ByteString.Lazy as B
import Data.Aeson
import Data.Text.Encoding

data Pos = Pos !Double !Double deriving (Eq, Ord, Show)

data Problem = Problem { roomWidth, roomHeight, stageWidth, stageHeight :: !Double
                       , stageBottomLeft :: [Double]
                       , musicians :: [Int]
                       , attendees :: [Attendee]
                       , pillars :: [Pillar]
                       } deriving (Show)

instance FromJSON Problem where
  parseJSON = withObject "Problem" $ \v -> Problem
      <$> v .: "room_width"
      <*> v .: "room_height"
      <*> v .: "stage_width"
      <*> v .: "stage_height"
      <*> v .: "stage_bottom_left"
      <*> v .: "musicians"
      <*> v .: "attendees"
      <*> v .: "pillars"

data Attendee = Attendee { x, y :: !Double, tastes :: [Double] } deriving (Show)

instance FromJSON Attendee where
  parseJSON = withObject "Attendee" $ \v -> Attendee
      <$> v .: "x"
      <*> v .: "y"
      <*> v .: "tastes"

data Pillar = Pillar { center :: [Double], radius :: !Double } deriving (Show)

instance FromJSON Pillar where
  parseJSON = withObject "Pillar" $ \v -> Pillar
      <$> v .: "center"
      <*> v .: "radius"

instance FromJSON Pos where
  parseJSON = withObject "Position" $ \v -> Pos <$> v .: "x" <*> v .: "y"

instance ToJSON Pos where
  toJSON (Pos x y) = object [ "x" .= x, "y" .= y ]
  toEncoding (Pos x y) = pairs ("x" .= x <> "y" .= y)

data Solution = Solution { placements :: [Pos] } deriving (Show)

instance FromJSON Solution where
  parseJSON = withObject "Solution" $ \v -> Solution <$> v .: "placements"

instance ToJSON Solution where
  toJSON s = object [ "placements" .= placements s ]
  toEncoding s = pairs ("placements" .= placements s)

{-
instance ToJSON Problem where
  toJSON (Problem{id_, data_}) =
    object [ "id" .= id_, "data" .= data_ ]
    
  toEncoding (Problem{id_, data_}) =
    pairs ("id" .= id_ <> "data" .= data_)
-}

readProblem :: FilePath -> IO Problem
readProblem f = do bs <- B.readFile f
                   case eitherDecode bs of
                     Right (String s) -> case eitherDecode (B.fromStrict (encodeUtf8 s)) of
                                           Right p -> return p
                                           Left err -> error ("Could not decode problem: " ++ err)
                     Right other -> error ("Cound not decode problem. Unexpected value: " ++ show other)
                     Left err -> error ("Could not decode problem: " ++ err) 

readSolution :: FilePath -> IO Solution
readSolution f = do bs <- B.readFile f
                    case eitherDecode bs of
                      Right s -> return s
                      Left err -> error ("Could not decode solution: " ++ err)